File: python/src/dag.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from typing import Deque, Generic, Iterable, Iterator, List, Set, Tuple, TypeVar

from exception import IllegalGraphException

T = TypeVar("T")


def validate_node(node_id: int, size: int) -> None:
    """Ensures ``node_id`` is valid for a graph of ``size`` nodes.

    Raises:
        IndexError: if out of range
    """
    if not 0 <= node_id < size:
        raise IndexError(f"Invalid node id: {node_id}")


@dataclass(frozen=True)
class Dag(Generic[T], Iterable[T]):
    """Immutable directed graph of nodes with integer ids and typed payloads.

    Obtain instances via the static :meth:`builder` factory; edges go from source to target.
    """

    _nodes: Tuple[T, ...]
    _adj: Tuple[Tuple[int, ...], ...]
    _rev: Tuple[Tuple[int, ...], ...]

    def __len__(self) -> int:
        """Number of nodes (ids are ``0 .. size()-1``)."""
        return len(self._nodes)

    def size(self) -> int:
        """Number of nodes."""
        return len(self)

    def __getitem__(self, node_id: int) -> T:
        """Payload for node ``node_id``."""
        validate_node(node_id, len(self))
        return self._nodes[node_id]

    def get(self, node_id: int) -> T:
        """Payload for node ``node_id``."""
        return self[node_id]

    def in_degree(self, node_id: int) -> int:
        """Count of incoming edges to ``node_id``."""
        validate_node(node_id, len(self))
        return len(self._rev[node_id])

    def out_degree(self, node_id: int) -> int:
        """Count of outgoing edges from ``node_id``."""
        validate_node(node_id, len(self))
        return len(self._adj[node_id])

    def targets(self, node_id: int) -> Iterable[int]:
        """Successor ids of ``node_id``."""
        validate_node(node_id, len(self))
        return self._adj[node_id]

    def sources(self, node_id: int) -> Iterable[int]:
        """Predecessor ids of ``node_id``."""
        validate_node(node_id, len(self))
        return self._rev[node_id]

    def __iter__(self) -> Iterator[T]:
        """Payloads in id order; use ``targets`` / ``sources`` for edge endpoints."""
        return iter(self._nodes)

    @staticmethod
    def validate_node(node_id: int, size: int) -> None:
        """Ensures ``node_id`` is valid for a graph of ``size`` nodes.

        Raises:
            IndexError: if out of range
        """
        validate_node(node_id, size)

    @classmethod
    def builder(cls) -> DagBuilder[T]:
        """Returns a new mutable ``Builder``."""
        return DagBuilder()


class DagBuilder(Generic[T]):
    """Mutable graph builder."""

    def __init__(self) -> None:
        self._nodes: List[T] = []
        self._adj: List[Set[int]] = []
        self._rev: List[Set[int]] = []

    def add(self, data: T) -> int:
        """Adds a node; returns its id for passing to ``connect``."""
        node_id = len(self._nodes)
        self._nodes.append(data)
        self._adj.append(set())
        self._rev.append(set())
        return node_id

    def connect(self, source: int, target: int) -> DagBuilder[T]:
        """Directed edge ``source`` → ``target`` (ignored if duplicate).

        Raises:
            IllegalGraphException: if ``source == target``
        """
        validate_node(source, len(self._nodes))
        validate_node(target, len(self._nodes))
        
        if source == target:
            raise IllegalGraphException(f"Self-loop not allowed: {source}")
            
        if target not in self._adj[source]:
            self._adj[source].add(target)
            self._rev[target].add(source)
        return self

    def build(self) -> Dag[T]:
        """Build the graph."""
        self._cycle_check()
        return Dag(
            _nodes=tuple(self._nodes),
            _adj=tuple(tuple(sorted(s)) for s in self._adj),
            _rev=tuple(tuple(sorted(s)) for s in self._rev),
        )

    def _cycle_check(self) -> None:
        n = len(self._nodes)
        if n == 0:
            return
            
        indeg = [len(self._rev[i]) for i in range(n)]
        ready: Deque[int] = deque(i for i, d in enumerate(indeg) if d == 0)
        processed = 0
        
        while ready:
            u = ready.popleft()
            processed += 1
            for v in self._adj[u]:
                indeg[v] -= 1
                if indeg[v] == 0:
                    ready.append(v)
                    
        if processed != n:
            raise IllegalGraphException("Graph contains a directed cycle")

File: typescript/src/dag.ts
import { IllegalGraphException } from "./illegalGraphException.js";

type Node<T> = { id: number; data: T };

/**
 * Immutable directed graph of nodes with integer ids and typed payloads. Obtain instances via the
 * static `Dag.builder()` factory; edges go from source to target.
 *
 * @typeParam T payload at each node
 */
export class Dag<T> implements Iterable<T> {
  readonly #nodes: readonly Node<T>[];
  readonly #adjacency: readonly ReadonlySet<number>[];
  readonly #reverseAdjacency: readonly ReadonlySet<number>[];

  private constructor(
    nodes: readonly Node<T>[],
    adjacency: readonly ReadonlySet<number>[],
    reverseAdjacency: readonly ReadonlySet<number>[],
  ) {
    this.#nodes = nodes;
    this.#adjacency = adjacency;
    this.#reverseAdjacency = reverseAdjacency;
  }

  /** Number of nodes (ids are `0 .. size-1`). */
  get size(): number {
    return this.#nodes.length;
  }

  /** Payload for node `id`. */
  get(id: number): T {
    Dag.validateNode(id, this.#nodes.length);
    return this.#nodes[id]!.data;
  }

  /** Count of incoming edges to `id`. */
  inDegree(id: number): number {
    Dag.validateNode(id, this.#nodes.length);
    return this.#reverseAdjacency[id]!.size;
  }

  /** Count of outgoing edges from `id`. */
  outDegree(id: number): number {
    Dag.validateNode(id, this.#nodes.length);
    return this.#adjacency[id]!.size;
  }

  /** Successor ids of `id`. */
  targets(id: number): readonly number[] {
    Dag.validateNode(id, this.#nodes.length);
    return Array.from(this.#adjacency[id]!);
  }

  /** Predecessor ids of `id`. */
  sources(id: number): readonly number[] {
    Dag.validateNode(id, this.#nodes.length);
    return Array.from(this.#reverseAdjacency[id]!);
  }

  /** Returns a new mutable `Builder`. */
  static builder<T>(): Dag.Builder<T> {
    return new Dag.Builder<T>();
  }

  /**
   * Ensures `id` is valid for a graph of `size` nodes.
   *
   * @throws RangeError if out of range
   */
  static validateNode(id: number, size: number): void {
    if (!Number.isInteger(id)) {
      throw new RangeError(`Invalid node id: ${String(id)}`);
    }
    if (id < 0 || id >= size) {
      throw new RangeError(`Invalid node id: ${id}`);
    }
  }

  /** Payloads in id order; use `targets` / `sources` for edge endpoints. */
  *[Symbol.iterator](): Iterator<T> {
    for (const node of this.#nodes) {
      yield node.data;
    }
  }
}

export namespace Dag {
  /**
   * Mutable graph builder.
   *
   * @typeParam T node payload type
   */
  export class Builder<T> {
    readonly #nodes: Node<T>[] = [];
    readonly #adjacency: Set<number>[] = [];
    readonly #reverseAdjacency: Set<number>[] = [];

    /** Adds a node; returns its id for passing to `connect`. */
    add(data: T): number {
      const id = this.#nodes.length;
      this.#nodes.push({ id, data });
      this.#adjacency.push(new Set());
      this.#reverseAdjacency.push(new Set());
      return id;
    }

    /**
     * Directed edge `source` → `target` (ignored if duplicate).
     *
     * @throws IllegalGraphException if `source === target`
     */
    connect(source: number, target: number): this {
      Dag.validateNode(source, this.#nodes.length);
      Dag.validateNode(target, this.#nodes.length);

      if (source === target) {
        throw new IllegalGraphException(`Self-loop not allowed: ${source}`);
      }

      const edges = this.#adjacency[source]!;
      if (!edges.has(target)) {
        edges.add(target);
        this.#reverseAdjacency[target]!.add(source);
      }
      return this;
    }

    /** Build the graph. */
    build(): Dag<T> {
      this.#cycleCheck();

      const nodes = Object.freeze([...this.#nodes]);
      const adjacency = Object.freeze(this.#adjacency.map((s) => Object.freeze(new Set(s))));
      const reverseAdjacency = Object.freeze(
        this.#reverseAdjacency.map((s) => Object.freeze(new Set(s))),
      );

      return new (Dag as any)(nodes, adjacency, reverseAdjacency);
    }

    #cycleCheck(): void {
      const n = this.#nodes.length;
      if (n === 0) return;

      const inDegree = this.#reverseAdjacency.map((s) => s.size);
      const ready: number[] = [];

      for (let i = 0; i < n; i++) {
        if (inDegree[i] === 0) ready.push(i);
      }

      let processed = 0;
      while (ready.length > 0) {
        const u = ready.shift()!;
        processed++;
        for (const v of this.#adjacency[u]!) {
          inDegree[v]--;
          if (inDegree[v] === 0) ready.push(v);
        }
      }

      if (processed !== n) {
        throw new IllegalGraphException("Graph contains a directed cycle");
      }
    }
  }
}
